Welding animation port

Vexylius opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Ports https://github.com/pariahstation/Pariah-Station/pull/665
but makes self-welding ipcs have the animation too
(doesn't work when somebody else welds them because it's instant)

looks something like this:
![68747470733a2f2f6964622e616666656374656461726330372e636f2e756b2f46696c65732f3336373235323037352f363434333136313335342f312f61667465722e676966](https://user-images.githubusercontent.com/22380218/168496125-eac93cec-779c-4cdd-9ad1-48a5d584b1a7.gif)


## Why It's Good For The Game
Makes welding pretty!
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Welding animation from TauCeti
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Vexylius committed:
Adds welding animation

Fikou reviewed (commented):
please post a video of you welding some weldable objects (objects that arent walls too)
also add this to plasma cutters

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -122,12 +127,15 @@
+					H.cut_overlay(GLOB.welding_sparks)
 					return
 			item_heal_robotic(H, user, 15, 0)
+			H.cut_overlay(GLOB.welding_sparks)
```
what youre doing here seems confusing. why not just use use_tool for the fixing chassis?
you can replace the use_tool being used for fuel consumption with use() and tool_start_check()

Vexylius replied:
Are you suggesting for me to rewrite the code that was previously written because I'm implementing a new feature in it?

Fikou replied:
yeah you are editing a proc in a way that can be done better with a tiny rewrite so im asking you to do so

Vexylius replied:
ah, sure, was just sitting for a while trying to figure out how am I supposed to do that without doing that, nvm lol

Vexylius replied:
Changed the code a bit.
I use use_tool because it does the delay, sound and vfx. I don't see a reason to use use instead of use_tool, as I would have to implement a few things that use_tool does in order to have the same functionality of that tool on usage.

Vexylius committed:
plasma cutter welding effect

Vexylius commented:
>please post a video of you welding some weldable objects (objects that arent walls too)
also add this to plasma cutters

https://user-images.githubusercontent.com/22380218/168925012-54ca745b-f225-41ef-9c5b-494ea5078e29.mp4

Vexylius committed:
Update weldingtool.dm

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,24 @@
+		user.visible_message("<span class='notice'>[user] starts to fix some of the dents on [H]'s [parse_zone(affecting.body_zone)].</span>",
+			"<span class='notice'>You start fixing some of the dents on [H == user ? "your" : "[H]'s"] [parse_zone(affecting.body_zone)].</span>")
+		use_tool(H, user, 50, 0, 25)
+		item_heal_robotic(H, user, 15, 0)
```
you will be able to heal instantly by canceling your do_afters. you need to return if use_tool fails.
you also nerfed it, it no longer heals instantly when someone else is doing the healing

Fikou replied:
use named arguments for the brute and burn heal

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,24 @@
+	. = ..()
+	target.cut_overlay(GLOB.welding_sparks)
+
 /obj/item/weldingtool/attack(mob/living/carbon/human/H, mob/user)
```
use a name that's not H, we want readability

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,24 @@
 	if(!istype(H))
 		return ..()
 
 	var/obj/item/bodypart/affecting = H.get_bodypart(check_zone(user.zone_selected))
```
affecting is also not a very readable namee

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,24 @@
-				if(!do_mob(user, H, 50))
-					return
-			item_heal_robotic(H, user, 15, 0)
+		if(!src.tool_start_check(user, amount=1))
```
don't do src.proc

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,24 @@
+			return
+		user.visible_message("<span class='notice'>[user] starts to fix some of the dents on [H]'s [parse_zone(affecting.body_zone)].</span>",
+			"<span class='notice'>You start fixing some of the dents on [H == user ? "your" : "[H]'s"] [parse_zone(affecting.body_zone)].</span>")
+		use_tool(H, user, 50, 0, 25)
```
use named arguments here, use a ternary operator for if you or someone else is doing the fixing, to set the time between 0 SECONDS and 5 SECONDS
you also want to set amount so it actually takes some fuel

Vexylius committed:
Horrible change to weld-fixing robo-limbs!

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,30 @@
-				if(!do_mob(user, H, 50))
-					return
-			item_heal_robotic(H, user, 15, 0)
+	if(attackedLimb && (!IS_ORGANIC_LIMB(attackedLimb)) && user.a_intent != INTENT_HARM)
```
you can reverse this statement and early return, and then unindent the stuff beneath the if

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,30 @@
+				return
+		else
+			if(!use_tool(target, user, delay=5, amount=0, volume=25))
+				return
```
just use && here, no need to make another "step" if contains nothing
also you can just use a tertiary statement here like i told you

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,30 @@
+	if(attackedLimb && (!IS_ORGANIC_LIMB(attackedLimb)) && user.a_intent != INTENT_HARM)
+		if(!tool_start_check(user, amount=1))
+			return
+		var/selfFix=(target == user)
```
```suggestion
		var/selfFix = (target == user)
```

Fikou commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -114,20 +114,30 @@
+		user.visible_message("<span class='notice'>[user] starts to fix some of the dents on [target]'s [parse_zone(attackedLimb.body_zone)].</span>",
+			"<span class='notice'>You start fixing some of the dents on [selfFix ? "your" : "[target]'s"] [parse_zone(attackedLimb.body_zone)].</span>")
+		if(selfFix)
+			if(!use_tool(target, user, delay=50, amount=0, volume=30)) //Setting this to 0 because afterattack seems to be handling fuel consumption already. No need to take the fuel twice.
```
cant you just return true to cancel afterattack?

Vexylius committed:
oh my god it works

Vexylius committed:
Fikou code improvement

Co-authored-by: Fikou <23585223+Fikou@users.noreply.github.com>

Latentish commented:
There's a comment on the PR this ports.
![image](https://user-images.githubusercontent.com/55075747/171063093-38dd0814-329a-4692-bd6a-d16d201ffe69.png)
So in summary, good PR, muh welding is a good precedent for other things to have animations as well. Idk what this guy is talking about.

rye-rice commented:
making animations is extremely time consuming if you dont have pixel fx like the original spriter made this with

Latentish commented:
This is extremely true, but this animation is also extremely good.

Fikou commented:
waiting for you to do the things i showed you on discord a week ago

ZephyrTFA committed:
Merge branch 'master' into welder-shmelder

triplezeta commented:
we're still waiting 🥺 

Aside from code review this PR has been TMed for awhile and seems fine.

Vexylius committed:
Fikou code rewrite

https://discord.com/channels/837744059291533392/837744059291533395/977684306824683550

Vexylius commented:
worst contributor 2003
